dora/core/common/src/test/java/alluxio/UfsUrlTest.java

```diff
@@ -118,13 +136,352 @@ public void hashCodeTest() {
+    assertFalse(ufsUrl2.isAncestorOf(ufsUrl1));
+    assertFalse(ufsUrl4.isAncestorOf(ufsUrl2));
+    assertFalse(ufsUrl3.isAncestorOf(ufsUrl6));
+    assertTrue(ufsUrl5.isAncestorOf(ufsUrl1));
+  }
+
+  /**
+   * Tests the {@link UfsUrl#join(String)}.
+   */
+  @Test
+  public void join() {
+    assertEquals(UfsUrl.createInstance("abc:/a"), UfsUrl.createInstance("abc:/").join("a"));
+    assertEquals(UfsUrl.createInstance("abc:/a/b.txt"),
+        UfsUrl.createInstance("abc:/a").join("/b.txt"));
+
+    // join empty string
+    assertEquals(UfsUrl.createInstance("abc:/a"), UfsUrl.createInstance("abc:/a").join(""));
+  }
+
+  /**
+   * Tests the {@link UfsUrl#join(String)} with special character.
+   */
+  @Test
+  public void joinSpecialChar() {
+    final String pathWithSpecialChar = "根目录";
+    assertEquals(UfsUrl.createInstance("abc:/" + pathWithSpecialChar),
+        UfsUrl.createInstance("abc:/").join(pathWithSpecialChar));
+
+    final String pathWithSpecialCharAndColon = "根目录";
+    assertEquals(UfsUrl.createInstance("abc:/" + pathWithSpecialCharAndColon),
+        UfsUrl.createInstance("abc:/").join(pathWithSpecialCharAndColon));
+  }
+
+  /**
+   * Tests the {@link UfsUrl#createInstance(String)} constructor to work with file URIs
+   * appropriately.
+   */
+  @Test
+  public void localFileUrl() {
+    UfsUrl url = UfsUrl.createInstance("file:///foo/bar");
+    assertTrue(url.getAuthority().get().toString().isEmpty());
+    assertEquals("/foo/bar", url.getFullPath());
+    assertEquals("file:///foo/bar", url.toString());
+  }
+
+  /**
+   * Tests the {@link UfsUrl#toString()} method to normalize paths.
+   */
+  @Test
+  public void normalize() {
+    assertEquals("xyz:///", UfsUrl.createInstance("xyz://").toString());
+    assertEquals("xyz:///foo", UfsUrl.createInstance("xyz:/foo/").toString());
+    assertEquals("xyz:///foo", UfsUrl.createInstance("xyz:/foo//").toString());
+    assertEquals("xyz:///foo/bar", UfsUrl.createInstance("xyz:/foo//bar").toString());
+
+    assertEquals("xyz:///foo/bar/boo", UfsUrl.createInstance("xyz:/foo/bar/boo").toString());
+    assertEquals("xyz:///foo/boo/baz",
+        UfsUrl.createInstance("xyz:/foo/////boo/baz").toString());
+
+    assertEquals("foo://bar boo:8080/abc/c",
+        UfsUrl.createInstance("foo://bar boo:8080/abc///c").toString());
+  }
+
+  /**
+   * Tests the {@link UfsUrl#UfsUrl(String, String, String)} constructor to throw an
+   * exception in case an empty path was provided.
+   */
+  @Test
+  public void constructFromEmptyPathTest2() {
+    assertThrows(NullPointerException.class, () -> {
+      new UfsUrl(null, null, null);
+    });
+  }
+
+  /**
+   * Tests the {@link UfsUrl#UfsUrl(String, String, String)} constructor to throw an
+   * exception in case an empty path was provided.
+   */
+  @Test
+  public void constructFromEmptyPathTest3() {
+    assertThrows(NullPointerException.class, () -> {
+      new UfsUrl("file", null, "");
+    });
+  }
```

Comment:
You may:
1. Either, collapse all these 3 `constructFromEmptyPathTest` cases into one `@Test`
2. Or, rename these 3 like, `nullPath`, `nullAuthority`, `nullScheme` etc